### DISCUSSION TITLE
Possibility to use both Julia and Python in a quarto document without pycall or knitr or reticulate

### DISCUSSION THREAD

**Discussion by Xinenomine** [quarto-dev/quarto-cli#5380] [Q&A]
Hello everyone, 
I looked everywhere and couldn't find what I was looking for. 
I would like to know if there is a way to actually use Julia + Python within the same quarto document. Something like: 

header: 
jupyter: [Julia-1.9, python3]
or kernel: [Julia-1.9, python3] something like this

 '''{python}
python code
'''

'''{julia}
julia code
'''

without using knitr or pycall or something like this. Right now, you can use either the python kernel OR the IJulia or else right ?
What do you do when you need to call in both python and julia ? 
 I wonder why It is so hard since jupyter should be able to be used for both Julia and python. I would like to be able to code seemlessly in python and julia like some sort of orgmode. Is it possible ? Will it be implemented ? 

I would like to thank you all for being part of this community. I thank the developpers of quarto for the great work !

**Comment by mcanouil** [comment]
It's one kernel/engine at a time.
If you are not willing to use `knitr`, `PyCall` or `PyJulia` to allow languages to talk to each other within the same engine/kernel, then you won't be able to use multiple languages.

Note that, at least for R/knitr as the kernel/engine, you can write:

````qmd
```{r}
<code>
```

```{python}
<code>
```
````

Behind the hood, Python code will actually be called/evaluated from R using `reticulate`.

As for Python/Jupyter kernels abilities, it should be the same way as for Python/Julia, better to check the kernel documentation to know more on this.
Note that there are a lot of possible kernels for Jupyter.

PS: what's the issue of having `reticulate`, `rpy2`, `PyCall` or `PyJulia` behind the scene?

**Comment by cderv** [comment]
> I wonder why It is so hard since jupyter should be able to be used for both Julia and python.

As discussed in the issue related to this question (https://github.com/quarto-dev/quarto-cli/issues/135#issuecomment-942383414), Jupyter kernel are single language (unless it changed). As mentioned above, Quarto use one kernel per document - so this is the limitation. 

Did you had other solution in mind ? How do you use Jupyter with several language in same notebook ?

**Reply by fkohrt** [reply]
While Jupyter only supports one kernel per notebook, there are kernels that allow polyglot execution, specifically [allthekernels](https://github.com/quarto-dev/quarto-cli/issues/3459) and [SoS](https://github.com/jupyterlab/jupyterlab/issues/2815#issuecomment-349773751). Only the latter appears to be maintained, though.

**Reply by cderv** [reply]
Thanks for helping cross reference topic ! 👍

**Comment by Xinenomine** [comment]
Thank you very much for your comments on this. I really appreciate it. 
Sos seems to be a great choice, I know there is PyJulia and co but I also like to have the ability to use different languages in the same quarto document without having to rely on these tweaks.
I tried Sos in my quarto document but I couldn't make it work though. Below is what I tried so far: 

```
---
title: "using Julia and Python with Sos"
format:
  html:
    code-fold: true
kernel: SoS
---

```{python}
#| label: fig-polar
#| fig-cap: "A line plot on a polar axis"

import numpy as np
import matplotlib.pyplot as plt

r = np.arange(0, 2, 0.01)
theta = 2 * np.pi * r
fig, ax = plt.subplots(
  subplot_kw = {'projection': 'polar'} 
)
ax.plot(theta, r)
ax.set_rticks([0.5, 1, 1.5, 2])
ax.grid(True)
plt.show()
```

```{julia}
#| label: fig-parametric
#| fig-cap: "Parametric Plots"

using Plots

plot(sin, 
     x->sin(2x), 
     0, 
     2π, 
     leg=false, 
     fill=(0,:lavender))
```

But it does not produce the Julia part as expected

I installed the sos kernel following the installation page on the Sos webpage though. 

Thank you very much for your help !

**Reply by fkohrt** [reply]
Sure, I can provide an example :)

After [installing the kernel](https://vatlab.github.io/sos-docs/running.html#Local-installation) and following the [instructions for every language](https://vatlab.github.io/sos-docs/running.html#Supported-Languages) one would like to use, the [documentation mentions the `%use` magic](https://vatlab.github.io/sos-docs/doc/user_guide/multi_kernel_notebook.html#Switch-between-kernels-), which is the key for doing this with Quarto.

<details>
<summary>Example</summary>

_Note that you need [palmer-penguins.csv](https://github.com/quarto-dev/quarto-web/blob/78d61df/docs/interactive/ojs/palmer-penguins.csv) for the Observable code to work._

``````qmd
---
title: "Polyglot Jupyter"
engine: jupyter
jupyter: sos
editor: source
---

# R

```{sos}
#| label: fig-airquality
#| fig-cap: "Temperature and ozone level."
#| warning: false
%use R

library(ggplot2)

ggplot(airquality, aes(Temp, Ozone)) + 
  geom_point() + 
  geom_smooth(method = "loess")
```

# Python

```{sos}
#| label: fig-polar
#| fig-cap: "A line plot on a polar axis"
%use SoS

import numpy as np
import matplotlib.pyplot as plt

r = np.arange(0, 2, 0.01)
theta = 2 * np.pi * r
fig, ax = plt.subplots(
  subplot_kw = {'projection': 'polar'} 
)
ax.plot(theta, r)
ax.set_rticks([0.5, 1, 1.5, 2])
ax.grid(True)
plt.show()
```

# Julia

```{sos}
#| label: fig-parametric
#| fig-cap: "Parametric Plots"
%use Julia

using Plots

plot(sin, 
     x->sin(2x), 
     0, 
     2π, 
     leg=false, 
     fill=(0,:lavender))
```

# Oberservable

```{ojs}
//| echo: false

//| panel: input
viewof bill_length_min = Inputs.range(
  [32, 50], 
  {value: 35, step: 1, label: "Bill length (min):"}
)
viewof islands = Inputs.checkbox(
  ["Torgersen", "Biscoe", "Dream"], 
  { value: ["Torgersen", "Biscoe"], 
    label: "Islands:"
  }
)
```

::: {.panel-tabset}

## Plot

```{ojs}
//| echo: false

Plot.rectY(filtered, 
  Plot.binX(
    {y: "count"}, 
    {x: "body_mass_g", fill: "species", thresholds: 20}
  ))
  .plot({
    facet: {
      data: filtered,
      x: "sex",
      y: "species",
      marginRight: 80
    },
    marks: [
      Plot.frame(),
    ]
  }
)
```

## Data

```{ojs}
//| echo: false

Inputs.table(filtered)
```

:::

```{ojs}
//| echo: false

data = FileAttachment("palmer-penguins.csv").csv({ typed: true })
```

```{ojs}
//| echo: false

filtered = data.filter(function(penguin) {
  return bill_length_min < penguin.bill_length_mm &&
         islands.includes(penguin.island);
})
```
``````

</details>

**Reply by cderv** [reply]
Thanks for sharing this ! @fkohrt Can I reuse your example for adding to our doc or quarto examples somewhere ?

**Reply by fkohrt** [reply]
Sure :)

Another feature that comes in handy for polyglot notebooks is [variable exchange among subkernels](https://vatlab.github.io/sos-docs/doc/user_guide/multi_kernel_notebook.html#Variable-exchange-among-subkernels). I haven't found a way to enable proper syntax highlighting though.

For the sake of completeness, here's a polyglot notebook with knitr as engine:

<details>
<summary>Example</summary>

_Note that you need [palmer-penguins.csv](https://github.com/quarto-dev/quarto-web/blob/78d61df/docs/interactive/ojs/palmer-penguins.csv) for the Observable code to work._

``````qmd
---
title: "Polyglot knitr"
engine: knitr
editor: source
---

# R

```{r}
#| label: fig-airquality
#| fig-cap: "Temperature and ozone level."
#| warning: false

library(ggplot2)

ggplot(airquality, aes(Temp, Ozone)) + 
  geom_point() + 
  geom_smooth(method = "loess")
```

# Python

```{python}
#| label: fig-polar
#| fig-cap: "A line plot on a polar axis"

import numpy as np
import matplotlib.pyplot as plt

r = np.arange(0, 2, 0.01)
theta = 2 * np.pi * r
fig, ax = plt.subplots(
  subplot_kw = {'projection': 'polar'} 
)
ax.plot(theta, r)
ax.set_rticks([0.5, 1, 1.5, 2])
ax.grid(True)
plt.show()
```

# Julia

JuliaCall has issues right now, see [Non-Contradiction/JuliaCall#206](https://github.com/Non-Contradiction/JuliaCall/issues/206).

```{julia}
#| label: fig-parametric
#| fig-cap: "Parametric Plots"
#| eval: false

using Plots

plot(sin, 
     x->sin(2x), 
     0, 
     2π, 
     leg=false, 
     fill=(0,:lavender))
```

# Oberservable

```{ojs}
//| echo: false

//| panel: input
viewof bill_length_min = Inputs.range(
  [32, 50], 
  {value: 35, step: 1, label: "Bill length (min):"}
)
viewof islands = Inputs.checkbox(
  ["Torgersen", "Biscoe", "Dream"], 
  { value: ["Torgersen", "Biscoe"], 
    label: "Islands:"
  }
)
```

::: {.panel-tabset}

## Plot

```{ojs}
//| echo: false

Plot.rectY(filtered, 
  Plot.binX(
    {y: "count"}, 
    {x: "body_mass_g", fill: "species", thresholds: 20}
  ))
  .plot({
    facet: {
      data: filtered,
      x: "sex",
      y: "species",
      marginRight: 80
    },
    marks: [
      Plot.frame(),
    ]
  }
)
```

## Data

```{ojs}
//| echo: false

Inputs.table(filtered)
```

:::

```{ojs}
//| echo: false

data = FileAttachment("palmer-penguins.csv").csv({ typed: true })
```

```{ojs}
//| echo: false

filtered = data.filter(function(penguin) {
  return bill_length_min < penguin.bill_length_mm &&
         islands.includes(penguin.island);
})
```
``````

</details>
